[week5] 고양이 장난감가게 만들기 (2)

SUBIN606 opened the pull request:
안녕하세요 이번 한 주도 잘 부탁드립니다 😀

## 과제 목표
- [x] 고양이 장난감가게 유효성 검사 추가하기
- [x] 회원 생성하기 - `POST /user`
- [x] 회원 수정하기 - `PATCH /user/{id}`
- [x] 회원 삭제하기 - `DELETE /user/{id}`

## 제한 조건
모든 기능은 TDD로 구현해야 합니다.

## 요구 사항
- [x] 테스트 커버리지 100%를 달성해야 합니다.
- [x] 모든 API 테스트를 통과해야 합니다.
- [x] 모든 E2E 테스트를 통과해야 합니다.

SUBIN606 committed:
initial commit

SUBIN606 commented:
4주차 과제 코드로 이어서 진행하도록 하겠습니다!

SUBIN606 committed:
test: 상품 등록 유효성 검사 테스트 작성

SUBIN606 committed:
feat: 상품 등록 시 유효성 검사

ProductDto에 validation 애노테이션을 등록해 상품 등록 요청 시 유효성 검사가 이루어지도록 합니다.

SUBIN606 committed:
test: 상품 수정 유효성 검사 테스트

SUBIN606 committed:
feat: 상품 수정 시 유효성 검사

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductUpdateServiceImpl.java:
```diff
@@ -0,0 +1,25 @@
+
+@RequiredArgsConstructor
+@Service
+public class ProductUpdateServiceImpl implements ProductUpdateService {
```
`updateService` 와 `saveService`가 나뉘어 있는 상황이 어색하다고 가정해 봅시다. 이런 경우에는 `updateServiceImpl`과 `saveServiceImpl`등을 하나로 합치는 걸 고려해 볼 수 있겠죠. `ProductCommandService` 클래스를 만들고, 이 클래스가 저장, 업데이트, 삭제 인터페이스를 모두 구현하는 방식으로 합쳐 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/ProductRepository.java:
```diff
@@ -1,14 +1,17 @@
+    List<Product> findAll();
 
-    Product save(Product product);
+    default Product save(ProductSaveRequest productSaveRequest) {
+        return save(productSaveRequest.product());
+    }
```
좋습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/ProductDeleteServiceTest.java:
```diff
@@ -0,0 +1,83 @@
+    @Nested
+    class Describe_delete_by_id {
+
+        @DisplayName("존재하는 상품 id가 주어진다면")
```
이제 이 표현을 좀 더 다듬어 봅시다. `존재하는`에 대해 생각해 볼 시간이 되었어요. 컴퓨터에서는 때때로 네트워크 단절이나, 정전, 느린 속도로 인한 타임아웃 등 다양한 문제로 `존재하는` 데이터를 접근하지 못하는 경우가 생길 수 있습니다. 그래서 많은 회사들이 `존재하지 않습니다`, `존재합니다` 같은 표현을 주의깊게 사용하고 있습니다. 여기에서는 어떤 표현을 생각할 수 있을까요?

SUBIN606 replied:
404 not found에서 착안하여 `찾은`, `찾을 수 없는`으로 변경해봤습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/ProductTest.java:
```diff
@@ -1,42 +1,71 @@
-        assertThat(product.getPrice()).isEqualTo(10000);
-        assertThat(product.getImageUrl())
-                .isEqualTo("http://localhost:8080/rat");
+    void getterTest() {
+        Product product
+                = new Product(NAME, MAKER, PRICE, URL);
+        assertAll(() -> {
+            assertThat(product.getId()).isEqualTo(null);
+            assertThat(product.getName()).isEqualTo(NAME);
+            assertThat(product.getMaker()).isEqualTo(MAKER);
+            assertThat(product.getPrice()).isEqualTo(PRICE);
+            assertThat(product.getImageUrl()).isEqualTo(URL);
+        });
```
이 테스트는 일종의 단순 노가다입니다. 정말 이 테스트가 필요한지에 대해 고민해 보세요.

SUBIN606 committed:
refactor: 리뷰 내용 적용

- ProductCommandService가 상품 조회 외 모든 역할을 담당합니다.
- 메서드명을 변경합니다.
- 테스트의 DisplayName을 변경합니다.

SUBIN606 committed:
test: Product getter 테스트 제거

SUBIN606 committed:
test: User 엔티티 테스트 작성

SUBIN606 committed:
feat: User 엔티티, 리포지토리 작성

사용자를 관리하기 위한 User 도메인을 작성합니다.

SUBIN606 committed:
test: 회원 생성 서비스 테스트 작성

SUBIN606 committed:
feat: 회원 생성 서비스 작성

- 사용자가 입력한 회원 정보를 전달할 DTO로 UserSaveDto를 만듭니다.
- 회원 생성에 필요한 메서드만 정의한 UserSaveRequest를 만듭니다.
- 회원 생성 및 변경 역할만 담당할 UserCommandService를 만듭니다.

SUBIN606 committed:
test: UserSaveController 테스트 작성

SUBIN606 committed:
feat: 회원 생성 컨트롤러

회원을 생성하는 역할을 담당하는 컨트롤러를 작성합니다.

johngrib reviewed (commented):
![image](https://user-images.githubusercontent.com/1855714/164002630-ae0e2e64-1b6f-4f56-ac11-148f351228bd.png)

커밋 메시지에 `feat:`, `test:` 같은 prefix를 꼭 붙이시는데요, 어떤 경우에는 바람직할 수 있지만 그렇지 않은 경우도 있습니다. 한 가지 방식에 손이 너무 익어버리지 않도록 다양한 방법으로 커밋 메시지를 작성하는 방법도 연습해 봅시다. 한동안 커밋을 할 때마다 다음 방법들을 랜덤하게 뽑아서 커밋 메시지를 작성해 보세요.

- 완성된 한국어 문장으로 작성해 봅니다.
- 반말로 작성해 봅니다. (`~했음` 도 괜찮습니다)
- 청유형으로 작성해 봅니다.
- 명령형으로 작성해 봅니다. (예: `상품변경 서비스를 테스트하라`)
- 작업을 잘 표시하는 단어 3~5개를 사용해 작성해 봅니다.
- 영어로 작성해 봅니다.

한 가지 방법에 익숙해지지 말고, 다양한 방법을 사용해보시면서 각 표현들의 장단점이 무엇인지를 고민해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductCommandService.java:
```diff
@@ -0,0 +1,45 @@
+
+
+/**
+ * 상품 변경 역할 담당
```
이렇게 해 봅시다.
```suggestion
 * 상품 변경을 담당합니다.
```

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductDeleteService.java:
```diff
@@ -0,0 +1,18 @@
+
+    /**
+     * 식별자에 해당하는 상품을 삭제합니다.
+     * 상품을 찾지 못할 경우 예외를 던집니다.
```
`@throws`가 있기 때문에 이 라인은 삭제해도 괜찮습니다.
```suggestion
```

johngrib commented on app/src/test/java/com/codesoom/assignment/controller/UserSaveControllerMockMvcTest.java:
```diff
@@ -0,0 +1,57 @@
+        repository.deleteAll();
+    }
+
+    @DisplayName("회원 정보를 성공적으로 저장한다.")
+    @Test
+    void saveUserTest() throws Exception {
+        final UserSaveDto userSaveDto = new UserSaveDto("홍길동", "email", "password");
+        final MvcResult result = mockMvc.perform(post("/users").accept(MediaType.APPLICATION_JSON_UTF8)
+                .content(objectMapper.writeValueAsString(userSaveDto))
+                .contentType(MediaType.APPLICATION_JSON))
+                .andExpect(status().isCreated())
+                .andReturn();
+
+        final UserResponseDto userResponseDto
+                = objectMapper.readValue(result.getResponse().getContentAsByteArray(), UserResponseDto.class);
+        assertThat(repository.findById(userResponseDto.getId())).isNotEmpty();
+    }
```
바람직하지 않은 경우나, 예외적인 경우도 테스트해보면 좋겠습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/controller/ProductUpdateControllerTest.java:
```diff
@@ -0,0 +1,90 @@
+            }
+        }
+
+        @DisplayName("id로 찾을 수 없는 상품 수정 요청이 오면")
```
이렇게 해 봅시다.
```suggestion
        @DisplayName("찾을 수 없는 상품 id가 주어지면")
```

johngrib commented on app/src/test/java/com/codesoom/assignment/controller/ProductReadControllerTest.java:
```diff
@@ -0,0 +1,102 @@
+    @Nested
+    class Describe_get_product {
+
+        @DisplayName("찾을 수 있는 상품 조회 요청이 오면")
+        @Nested
+        class Context_with_exist_id {
+
+            private Long EXIST_ID;
+
+            @BeforeEach
+            void setup() {
+                this.EXIST_ID = repository.save(SAVED_PRODUCT).getId();
+            }
+
+            @DisplayName("해당 상품을 반환한다.")
+            @Test
+            void will_return_found_product() {
+                assertThat(controller.getProduct(EXIST_ID)).isNotNull();
+            }
```
메소드 시그니처를 의식하고 `@DisplayName`을 작성해 보세요.
```suggestion
        @DisplayName("찾을 수 있는 상품의 id 가 주어지면")
        @Nested
        class Context_with_exist_id {

            private Long EXIST_ID;

            @BeforeEach
            void setup() {
                this.EXIST_ID = repository.save(SAVED_PRODUCT).getId();
            }

            @DisplayName("해당 상품을 반환한다.")
            @Test
            void will_return_found_product() {
                assertThat(controller.getProduct(EXIST_ID)).isNotNull();
            }
```

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserSaveTest.java:
```diff
@@ -0,0 +1,43 @@
+    @BeforeEach
+    void setup() {
+        this.service = new UserCommandService(repository);
+        repository.deleteAll();
```
중복이 있군요.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/ProductUpdateTest.java:
```diff
@@ -0,0 +1,95 @@
+            }
+        }
+
+        @DisplayName("id로 상품을 찾을 수 없다면")
```
이렇게 해 보세요.
`찾을 수 없는 상품 id가 주어지면`

SUBIN606 committed:
javaDoc을 작성합니다.

SUBIN606 committed:
회원 생성 실패 테스트

유효성 검증을 통과하지 못할 경우에 대한 테스트를 추가합니다.

SUBIN606 committed:
리뷰 내용을 적용합니다.

테스트 클래스의 DisplayName을 메서드에 맞게 수정합니다.
중복 메서드를 정리합니다.

SUBIN606 committed:
회원 정보 수정 테스트를 작성합니다.

SUBIN606 committed:
회원 정보 수정에 대한 기능을 추가합니다.

UserUpdateService가 회원 정보 수정을 담당합니다.
id로 회원을 찾지 못할 경우 UserNotFoundException을 던집니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controller/ProductControllerAdvice.java:
```diff
@@ -9,9 +9,9 @@
+ * 상품 컨트롤러에서 던지는 예외를 받아 응답합니다.
  */
-@RestControllerAdvice
+@RestControllerAdvice(annotations = ProductController.class)
```
커밋 메시지는 `javaDoc을 작성합니다.` 인데요, 커밋 메시지와 관계 없는 내용이 포함되어 있습니다. 이런 경우에는 코드 리뷰를 하는 것처럼 코멘트를 남겨 보세요. 예를들어 이렇게 할 수 있겠죠.

"이건 커밋 주제와 맞지 않는 변경이지만 git add를 할 때 실수로 포함시켜버렸습니다. 심각한 문제는 아니라 생각해서 작업한 그대로 올렸습니다. 양해 부탁드립니다."

작은 표현 하나로 매너있으면서, 실수를 인정하면서, 커밋 주제와 다른 코드를 리뷰어에게 먼저 안내하는 코멘트라 할 수 있겠습니다. 동료들끼리 많이 친해진 후에는 생략해도 좋겠지만 그렇기 전에는 이렇게 커뮤니케이션하면 더 좋겠죠.

![image](https://user-images.githubusercontent.com/1855714/164229963-562101b0-9296-48ac-9f1c-b9b03bddcc1f.png)

johngrib commented on app/src/test/java/com/codesoom/assignment/controller/UserSaveControllerMockMvcTest.java:
```diff
@@ -39,19 +40,85 @@ void cleanup() {
+            }
+        }
+
+        @DisplayName("이름을 입력하지 않으면")
+        @Nested
+        class Context_with_empty_name {
+
+            private final UserSaveDto EMPTY_NAME
+                    = new UserSaveDto("", "email", "password");
```
`이름을 입력하지 않으면`의 경우에 대해 더 다양한 테스트케이스를 고려할 수 있을 것 같습니다. `null`도 있고 `"  "`, `"\n"` 같은 것도 있겠죠.

johngrib reviewed (commented):
`리뷰 내용을 적용합니다`와 같은 커밋 메시지는 실제로 실무에서 많이 볼 수 있는 커밋 메시지입니다. 하지만 흔하게 본다고 해서 항상 좋은 것은 아닙니다. 연습할 때에는 좀 더 고민해보면서 더 나은 학습이 되도록 해보면 좋겠죠. `리뷰 내용을 적용한다`는 내용의 커밋 메시지를 코드숨 학습 기간 동안 사용 금지한다고 생각해 보세요. 그리고 어떤 커밋 메시지가 좋을지 3개 정도 후보 문장을 만들고 그 중에서 선택하는 연습을 해 보세요.

![image](https://user-images.githubusercontent.com/1855714/164230823-dd52d4be-4463-4de2-bdad-1d671332d40b.png)

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserUpdateTest.java:
```diff
@@ -0,0 +1,84 @@
+            @DisplayName("회원 정보를 수정 후 수정된 정보를 반환한다.")
+            @Test
+            void it_update_user() {
+                final UserSaveDto USER_TO_UPDATE = new UserSaveDto("임꺽정", "email", "password");
```
`USER_TO_UPDATE`도 Context 담당입니다. `@Test` 바깥에서 테스트를 준비하는 용도로 사용해 보세요.

SUBIN606 committed:
회원 생성 유효성 검사 케이스 추가

UserSaveControllerMockMvcTest에 유효성 검사 케이스를 추가합니다.

SUBIN606 committed:
테스트 컨텍스트에서 테스트를 준비하도록 수정

UserUpdateTest의 USER_TO_UPDATE를 Context 영역에서 준비하도록 수정합니다.

SUBIN606 committed:
회원 관련 요청에만 대응하는 ContollerAdvice 작성

회원 요청에만 대응하도록 UserControllerAdvice를 생성합니다.
회원을 찾지 못한 경우 던지는 예외 핸들링을 추가합니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/controller/UserSaveControllerMockMvcTest.java:
```diff
@@ -74,14 +76,23 @@ class Context_with_empty_name {
+                    = new UserSaveDto(" ", "email", "password");
+            private final UserSaveDto NULL_NAME
+                    = new UserSaveDto(null, "email", "password");
+            private final List<UserSaveDto> INVALID_USER_DTO = List.of(EMPTY_NAME, BLANK_NAME, NULL_NAME);
```
저도 종종 쓰는 방법입니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/controller/UserSaveControllerMockMvcTest.java:
```diff
@@ -74,14 +76,23 @@ class Context_with_empty_name {
-                        .contentType(MediaType.APPLICATION_JSON))
-                        .andExpect(status().isBadRequest());
+
+                for (UserSaveDto userSaveDto : INVALID_USER_DTO) {
+                    mockMvc.perform(post("/users").accept(MediaType.APPLICATION_JSON_UTF8)
+                            .content(objectMapper.writeValueAsString(userSaveDto))
+                            .contentType(MediaType.APPLICATION_JSON))
+                            .andExpect(status().isBadRequest());
+                }
```
좋습니다. 다만 변수 이름은 좀 더 낫게 고칠 수 있을 것 같군요.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserUpdateTest.java:
```diff
@@ -39,6 +39,8 @@ class Describe_update_user {
         class Context_with_exist_user {
             private Long EXIST_USER_ID;
+            private final UserSaveDto USER_TO_UPDATE
+                    = new UserSaveDto("임꺽정", "email", "password");
```
이메일과 패스워드를 좀 더 진짜같은 값으로 써보도록 합시다.

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -0,0 +1,36 @@
+        assertThat(new User()).isNotNull();
+    }
+
+    @DisplayName("생성자로 사용자 엔티티를 생성할 수 있다.")
```
생성자는 원래 생성하는 게 용도라서 이 설명은 좀 어색하군요. 더 나은 문구가 있을 겁니다.

SUBIN606 replied:
`new` 연산자로 객체를 인스턴스화 하고, 생성자의 인자 값 들로 객체를 초기화 할 수 있다는 것에서 착안하여 수정해봤습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/ProductTest.java:
```diff
@@ -1,42 +1,57 @@
+        assertThat(product.getName()).isEqualTo(NAME + "ver2");
+    }
+
+    @DisplayName("toString은 상품에 저장된 내용을 보여준다.")
```
`toString`의 리턴값 타입이 무엇인지 떠올리며 생각해 보세요. `toString`이 무엇을 보여준다는 것은 어떤 의미일까요?

johngrib commented on app/src/test/java/com/codesoom/assignment/controller/UserSaveControllerMockMvcTest.java:
```diff
@@ -0,0 +1,149 @@
+            private final UserSaveDto EMPTY_EMAIL
+                    = new UserSaveDto("홍길동", "", "password");
+            private final UserSaveDto BLANK_EMAIL
+                    = new UserSaveDto("홍길동",   "", "password");
```
이걸 의도하셨을 것 같습니다.
```suggestion
                    = new UserSaveDto("홍길동",  " ", "password");
```

SUBIN606 committed:
회원 변경 테스트를 수정하라

- 테스트에 사용되는 값들을 좀 더 실제와 비슷한 값으로 변경했습니다.
- Context에서 테스트에 이용되는 값을 준비하도록 코드를 이동합니다.
- 수정 후 값을 상수로 선언합니다.

SUBIN606 committed:
회원 생성 컨트롤러 테스트를 수정하라

- 오타를 수정했습니다.
- 변수명을 변경합니다.
- 이메일과 비밀번호 유효성 검사 케이스를 추가했습니다.

SUBIN606 committed:
회원 생성/요청 유효성 검사를 추가하라

- 이메일에 `@Email`어노테이션으로 이메일 형식 검증을 추가합니다.
- 비밀번호는 8~20자로 제한합니다.

SUBIN606 committed:
상품 엔티티 테스트의 DisplayName을 수정하라

- `toString()`테스트의 DisplayName을 수정합니다.

SUBIN606 committed:
회원 엔티티 테스트의 DisplayName을 수정하라

- 생성자로 객체를 인스턴스화 하고 값을 초기화 할 수 있다는 표현으로 수정합니다.

SUBIN606 committed:
회원 생성 컨트롤러 테스트를 수정하라

- 오타를 수정합니다.

SUBIN606 committed:
회원 수정 컨트롤러 테스트를 작성하라

- 회원 수정 요청 컨트롤러 테스트를 작성합니다.

SUBIN606 committed:
회원 수정 컨트롤러를 작성하라

- 회원 수정 요청을 처리하는 컨트롤러를 작성합니다.

SUBIN606 committed:
회원 삭제 서비스 테스트를 작성하라

- 회원 삭제 기능 테스트 코드를 먼저 작성합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductReadServiceImpl.java:
```diff
@@ -0,0 +1,28 @@
+
+    @Override
+    public Product findById(Long id) {
+        Product product = repository.findById(id).orElseThrow(() -> new ProductNotFoundException(id));
```
- 한 줄에 `.`이 하나 오도록 개행해 보세요.
- 예외를 생성할 때 `id`만 주지 말고 예외 메시지를 직접 작성해서 넣어 보세요.

SUBIN606 committed:
회원 삭제 서비스를 작성하라

- 회원 삭제를 담당하는 `UserDeleteService`인터페이스를 생성합니다.
- `UserCommandService`에서 회원 삭제를 구현합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductSaveRequest.java:
```diff
@@ -0,0 +1,24 @@
+
+    String getMaker();
+
+    BigDecimal getPrice();
```
당장 작업할 건 아니지만 고려해야 할 문제가 하나 있습니다. 만약 고양이 장난감 가게가 사업이 아주 잘 되어서 다른 나라로도 진출하게 되었다고 합시다. 그렇다면 이제부터는 통화 `currency`도 고려해야 할 수 있을 겁니다. 이런 상황에 대해 어떻게 대처해야 할 지 상상해 보는 시간을 가져 보세요.

SUBIN606 replied:
고민해봤는데요!

1. 상품을 등록할 때는 원화로 등록하고, 사용자가 접속한 나라 혹은 선택한 통화가 무엇인지 체크 후 원화 가격을 환산하여 보여준다. 하지만 이렇게 할 경우 판매자가 원하는 가격을 보여줄 수 없겠네요. 9900원과 10000원의 차이 같은..
2. `currency`필드를 추가해 상품 등록 시 통화에 맞는 값을 입력한다. 이렇게하면 판매자는 등록 시 판매를 원하는 나라를 선택할 수 있게 될 것 같습니다. 하지만 여러 나라에 판매할 상품을 등록할 땐 여러 번 등록을 해야 합니다.
3.  `Currency` enum클래스를 만들고, 대표적인 통화 (예: USD, EUR 등)를 먼저 정의 해둔 후 판매자가 각각 입력해 등록할 수 있도록 한다. 이렇게 하면.. 가격 필드가 엄청 많아질 수 있습니다.

잠깐 고민해본건데 참 어렵네요..ㅎㅎ

johngrib replied:
이런 문제는 정책 관리자와 토의해서 결정해야 하는 일이기도 합니다. 당장 구현하지 않더라도 이렇게 고민해보신 잠깐의 경험이 나중에 도움이 될 수도 있겠죠. 이 문제는 생각보다 까다로우면서 흔하게 접할 수 있는 문제라 꾸준히 생각하면서 어떻게 할지 고민해 보시면 좋을 거에요. 이와 관련해서는 마틴 파울러의 "엔터프라이즈 애플리케이션 아키텍처 패턴"이 읽어볼만한 책이라 할 수 있습니다. 돈을 프로그래밍에서 어떻게 다룰지에 대한 이야기가 나와요.

johngrib commented on app/src/test/java/com/codesoom/assignment/domain/UserTest.java:
```diff
@@ -0,0 +1,36 @@
+    private static final String EMAIL = "email";
+    private static final String PASSWORD = "password";
+
+    @DisplayName("기본 생성자는 회원 인스턴스를 만들 수 있다.")
```
"만들 수 있다"보다 더 강한 표현을 사용해 봅시다.
```suggestion
    @DisplayName("기본 생성자는 회원 인스턴스를 리턴한다.")
```

johngrib reviewed (approved):
전반적으로 코드가 깔끔하고 테스트 코드에도 정성을 많이 기울이셨습니다.

SUBIN606 committed:
회원 삭제 컨트롤러 테스트를 작성하라

- 회원 삭제 요청을 처리할 컨트롤러 테스트를 작성합니다.

SUBIN606 committed:
회원 삭제 요청 컨트롤러 기능을 구현하라

- DELETE 요청을 처리할 컨트롤러를 작성합니다.

SUBIN606 committed:
의존성을 변경합니다.

- 컨트롤러의 의존성을 구현체가 아닌 인터페이스로 변경합니다.

SUBIN606 committed:
ProductNotFoundException에 메시지를 전달하라

- 예외 메시지를 직접 작성해 예외를 던질 수 있습니다.

SUBIN606 committed:
회원 엔티티 테스트를 수정하라

- 기본 생성자 테스트의 `@DisplayName`을 수정합니다.

SUBIN606 committed:
스프링 부트 실행 테스트를 추가하라

- 스프링 부트 애플리케이션 실행 테스트를 추가합니다.

SUBIN606 committed:
User클래스의 의미 없는 코드를 제거하라

SUBIN606 committed:
api 테스트를 수정하라

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductReadServiceImpl.java:
```diff
@@ -21,7 +21,9 @@ public List<Product> findAll() {
-        Product product = repository.findById(id).orElseThrow(() -> new ProductNotFoundException(id));
+        Product product = repository.findById(id)
+                .orElseThrow(() ->
+                        new ProductNotFoundException("요청하신 상품을 찾지 못했습니다."));
```
이 예외 메시지를 읽게 되는 사람이 새벽 3시에 디버깅을 하고 있는 미래의 자기 자신이라고 상상해보세요. 아마 이런 메시지가 좀 더 미래의 자신에게 반가울 것입니다.

"id 3121 에 해당하는 상품을 찾지 못했습니다."

SUBIN606 committed:
예외 메시지를 수정하라

- 상품을 찾지 못했을 때 던질 예외 메시지에 id를 포함합니다.

johngrib commented:
한 주간 고생 많으셨습니다!
